### PR TITLE
docs: add design proposal for L4 HealthCheck

### PR DESCRIPTION
**Pull request by yskopets**
### Summary

* add design proposal for `L4 HealthCheck`

### Related issues

#393

### CODE REVIEW

File: docs/proposals/HealthChecks-L4.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+
+* Health xDS:
+  * as part of this API, Envoy receives assignments and reports results back to the Control Plane over gRPC
+    * see an example in the end
+  * is not considered production ready yet (developed in summer 2018 by an intern)
+  * does not support mTLS
+
+Conclusions:
+* we can already use `Health xDS` for `Envoy -> local app` health checks
+* changes to the Envoy will be necessary to use `Health xDS` for `Envoy -> upstream` "health checks" (add support for mTLS)
```
Comment (jakubdyszkiewicz) [inline docs/proposals/HealthChecks-L4.md]: What is the use case of using HDS for `Envoy -> upstream`?  Wouldn't it be better to only HC your local app and send status to CP, which then updates list of endpoints for dataplanes that use this service?

File: docs/proposals/HealthChecks-L4.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+* there are several ways how Envoy can be configured to obtain health statuses for upstream endpoints
+  * Envoy can do "active" health checks
+    * for TCP it means that Envoy will periodically try establishing a connection
+  * Envoy can do "passive" health checks
+    * for TCP it means that Envoy will be remembering statuses of recent connection attempts
+  * Envoy can receive health statuses from the Control Plane as part of EDS API
+
+## Requirements
+
+1. support `Envoy -> upstream` "health checks"
```
Comment (jakubdyszkiewicz) [inline docs/proposals/HealthChecks-L4.md]: I don't think this scales very well. At least for active health checks.
Let's say we've got app `backend` and 10 other apps with 10 instances each that communicate with it. HC is sent every second. Now we generate 100rps to `backend`.

File: docs/proposals/HealthChecks-L4.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+* Health xDS:
+  * as part of this API, Envoy receives assignments and reports results back to the Control Plane over gRPC
+    * see an example in the end
+  * is not considered production ready yet (developed in summer 2018 by an intern)
+  * does not support mTLS
+
+Conclusions:
+* we can already use `Health xDS` for `Envoy -> local app` health checks
+* changes to the Envoy will be necessary to use `Health xDS` for `Envoy -> upstream` "health checks" (add support for mTLS)
+* changes to the Envoy will be necessary to send event logs to the Control Plane (instead of logging to a file)
```
Comment (jakubdyszkiewicz) [inline docs/proposals/HealthChecks-L4.md]: Does it make sense to use active HC events when HDS is used?

For passive HC (outlier detection) I think this is "very local" for connection between A -> B. What would you do with information that B does not work from A perspective in the control plane?

File: docs/proposals/HealthChecks-L4.md (hunk 2 above)
Comment (yskopets) [reply docs/proposals/HealthChecks-L4.md]: It's always a user's choice.

Active health checks make perfect sense when your infrastructure is not big.

File: docs/proposals/HealthChecks-L4.md (hunk 1 above)
Comment (yskopets) [reply docs/proposals/HealthChecks-L4.md]: success of `Envoy -> local app` check doesn't give a full picture:
* it doesn't account for mTLS between client and server
* it doesn't account for different geographical location (e.g., connectivity to stand-by instances in another datacenter)

File: docs/proposals/HealthChecks-L4.md (hunk 3 above)
Comment (yskopets) [reply docs/proposals/HealthChecks-L4.md]: the goal of a Control Plane is to be smart and help users in every possible way.

E.g.,  
* make it visible that the problem is local to single dataplane
* make it visible that the problem is specific to certain geo location

**Pull request merged**